### PR TITLE
Add FoldNormal Distribution

### PR DESCRIPTION
**Pull request by CaptainMushroom**
Add FoldNormal Distribution

### CODE REVIEW

File: zhusuan/distributions/univariate.py [hunk 1]
```diff
@@ -163,6 +164,140 @@ def _log_prob(self, given):
...
 
+class HalfNormal(Distribution):
+    """
+    The class of univariate HalfNormal distribution.
+    See :class:`~zhusuan.distributions.base.Distribution` for details.
+
+    .. warning::
+
+         The order of arguments `logstd`/`std` will change to `std`/`logstd`
+         in the coming version.
```
Comment (thjashin) [inline zhusuan/distributions/univariate.py]: Just change it and remove this warning (in the doc string and in the code)

File: zhusuan/distributions/univariate.py [hunk 2]
```diff
@@ -163,6 +164,140 @@ def _log_prob(self, given):
...
+
+    :param mean: A `float` Tensor. The mean of the HalfNormal distribution.
+        The mean value must be 0.
+        Should be broadcastable to match `logstd`.
+    :param logstd: A `float` Tensor. The log standard deviation of the HalfNormal
+        distribution. Should be broadcastable to match `mean`.
+    :param std: A `float` Tensor. The standard deviation of the HalfNormal
+        distribution. Should be positive and broadcastable to match `mean`.
+    :param group_event_ndims: A 0-D `int32` Tensor representing the number of
```
Comment (thjashin) [inline zhusuan/distributions/univariate.py]: Now named group_ndims

File: zhusuan/distributions/univariate.py [hunk 3]
```diff
@@ -163,6 +164,140 @@ def _log_prob(self, given):
...
+    :param check_numerics: Bool. Whether to check numeric issues.
+    """
+
+    def __init__(self,
+                 mean=0.,
+                 logstd=None,
+                 std=None,
+                 group_event_ndims=0,
+                 is_reparameterized=True,
+                 check_numerics=False):
```
Comment (thjashin) [inline zhusuan/distributions/univariate.py]: check other distribution, here is a **kwargs passed to the base class.

File: zhusuan/distributions/univariate.py [hunk 4]
```diff
@@ -163,6 +164,140 @@ def _log_prob(self, given):
...
+        self._mean = tf.convert_to_tensor(mean)
+        warnings.warn("Normal: The order of arguments logstd/std will change "
+                      "to std/logstd in the coming version.")
+        if mean != 0.:
```
Comment (thjashin) [inline zhusuan/distributions/univariate.py]: just remove this parameter in \_\_init\_\_

Comment (thjashin) [reply zhusuan/distributions/univariate.py]: And you can not use if to check Tensors.

File: zhusuan/model/stochastic.py [hunk 5]
```diff
@@ -84,6 +85,54 @@ def __init__(self,
         )
         super(Normal, self).__init__(name, norm, n_samples)
 
+#TODO: changed
+class HalfNormal(StochasticTensor):
+    """
+    The class of univariate HalfNormal distribution.
```
Comment (thjashin) [inline zhusuan/model/stochastic.py]: The doc string should follow other stochastic tensors

File: zhusuan/model/stochastic.py [hunk 6]
```diff
@@ -84,6 +85,54 @@ def __init__(self,
...
+
+    def __init__(self,
+                 name,
+                 mean=0.,
+                 logstd=None,
+                 std=None,
+                 n_samples=None,
+                 group_event_ndims=0,
+                 is_reparameterized=True,
+                 check_numerics=False):
```
Comment (thjashin) [inline zhusuan/model/stochastic.py]: also **kwargs should be passed to base

File: zhusuan/distributions/univariate.py [hunk 7]
```diff
@@ -163,6 +164,140 @@ def _log_prob(self, given):
...
+    def _log_prob(self, given):
+        #TODO: Raise Error
+        c = -0.5 * np.log(2 * np.pi)
+        precision = tf.exp(-1 * self.logstd)
+        if self._check_numerics:
+            precision = tf.check_numerics(precision, "precision")
+        temp = given >= 0
+        temp = tf.cast(temp, tf.float32)
+        #return c - self.logstd - 0.5 * precision * tf.square(given - self.mean)
+        return (0.5 * (tf.log(2.0) - tf.log(np.pi)) - self.logstd - 0.5 * tf.square(given * tf.exp(-1 * self.logstd))) + tf.log(temp)
```
Comment (thjashin) [inline zhusuan/distributions/univariate.py]: a line space here

File: zhusuan/distributions/univariate.py [hunk 8]
```diff
@@ -163,6 +164,140 @@ def _log_prob(self, given):
...
+        return samples
+
+    def _log_prob(self, given):
+        #TODO: Raise Error
+        c = -0.5 * np.log(2 * np.pi)
+        precision = tf.exp(-1 * self.logstd)
+        if self._check_numerics:
+            precision = tf.check_numerics(precision, "precision")
+        temp = given >= 0
+        temp = tf.cast(temp, tf.float32)
```
Comment (thjashin) [inline zhusuan/distributions/univariate.py]: use tf.to_float

Comment (thjashin) [inline zhusuan/distributions/univariate.py]: rename temp, e.g., mask

File: zhusuan/distributions/univariate.py (hunk 7 above)
Comment (thjashin) [reply zhusuan/distributions/univariate.py]: use c and precision instead because you've defined them

File: zhusuan/distributions/univariate.py [hunk 9]
```diff
@@ -163,6 +164,140 @@ def _log_prob(self, given):
...
+        #return c - self.logstd - 0.5 * precision * tf.square(given - self.mean)
+        return (0.5 * (tf.log(2.0) - tf.log(np.pi)) - self.logstd - 0.5 * tf.square(given * tf.exp(-1 * self.logstd))) + tf.log(temp)
+    def _prob(self, given):
+        return tf.exp(self._log_prob(given))
+
```
Comment (thjashin) [inline zhusuan/distributions/univariate.py]: A test should be added in tests/distributions/test_univariate.py. Following other distributions.

File: zhusuan/model/stochastic.py (hunk 6 above)
Comment (thjashin) [reply zhusuan/model/stochastic.py]: A test should be added. Following other stochastictensors.

**Comment by thjashin** [comment]
THX for contributing. See the comments.

**Comment by CaptainMushroom** [comment]
Added checks

File: zhusuan/distributions/univariate.py [hunk 10]
```diff
@@ -39,12 +39,9 @@ class Normal(Distribution):
     """
     The class of univariate Normal distribution.
     See :class:`~zhusuan.distributions.base.Distribution` for details.
-
```
Comment (miskcoo) [inline zhusuan/distributions/univariate.py]: Don't remove the empty line, this will affect the docs.

File: zhusuan/distributions/univariate.py [hunk 11]
```diff
@@ -278,23 +267,22 @@ def _sample(self, n_samples):
...
+        precision = tf.exp(-2.0 * self.logstd)
         if self._check_numerics:
             precision = tf.check_numerics(precision, "precision")
-        temp = given >= 0
-        temp = tf.cast(temp, tf.float32)
-        #return c - self.logstd - 0.5 * precision * tf.square(given - self.mean)
-        return (0.5 * (tf.log(2.0) - tf.log(np.pi)) - self.logstd - 0.5 * tf.square(given * tf.exp(-1 * self.logstd))) + tf.log(temp)
+        mask = tf.log(tf.cast(given >= 0., dtype=precision.dtype))
+        return (c - (self.logstd + 0.5 * precision * tf.square(given - self.mean)) + \
+                tf.log(1.0 + tf.exp(-2.0 * self.mean * given * precision))) + mask
```
Comment (miskcoo) [inline zhusuan/distributions/univariate.py]: tf.nn.softplus may be better.

File: zhusuan/distributions/univariate.py (hunk 1 above)
Comment (miskcoo) [reply zhusuan/distributions/univariate.py]: maybe we need to keep the consistency between Normal and FoldNormal and not change the order since the two distributions are quite similar ? @thjashin

**Comment by CaptainMushroom** [comment]
Applied Some Suggestions

File: zhusuan/model/stochastic.py [hunk 12]
```diff
@@ -85,6 +86,56 @@ def __init__(self,
...
+        dimensions in `batch_shape` (counted from the end) that are grouped
+        into a single event, so that their probabilities are calculated
+        together. Default is 0, which means a single value is an event.
+        See :class:`~zhusuan.distributions.base.Distribution` for more detailed
+        explanation.
+    :param is_reparameterized: A Bool. If True, gradients on samples from this
+        distribution are allowed to propagate into inputs, using the
+        reparametrization trick from (Kingma, 2013).
+    :param check_numerics: Bool. Whether to check numeric issues.
+    """
```
Comment (thjashin) [inline zhusuan/model/stochastic.py]: docstring not fixed yet
